### ISSUE TITLE
Dynamic volume provisioning creates EBS volume in the wrong availability zone

### ISSUE THREAD

**Issue by jimmycuadra** [kubernetes/kubernetes#39178]
**What keywords did you search in Kubernetes issues before filing this one?** (If you have found any duplicates, you should instead reply there.): dynamic volume provisioning

---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one): bug report

**Kubernetes version** (use `kubectl version`):

```
Client Version: version.Info{Major:"1", Minor:"5", GitVersion:"v1.5.1", GitCommit:"82450d03cb057bab0950214ef122b67c83fb11df", GitTreeState:"clean", BuildDate:"2016-12-22T13:59:22Z", GoVersion:"go1.7.4", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"5", GitVersion:"v1.5.1", GitCommit:"82450d03cb057bab0950214ef122b67c83fb11df", GitTreeState:"clean", BuildDate:"2016-12-14T00:52:01Z", GoVersion:"go1.7.4", Compiler:"gc", Platform:"linux/amd64"}
```

**Environment**:
- **Cloud provider or hardware configuration**: AWS
- **OS** (e.g. from /etc/os-release): CoreOS 1185.5.0
- **Kernel** (e.g. `uname -a`): Linux ip-10-0-1-121.ec2.internal 4.7.3-coreos-r3 #1 SMP Wed Dec 7 09:29:55 UTC 2016 x86_64 Intel(R) Xeon(R) CPU E5-2676 v3 @ 2.40GHz GenuineIntel GNU/Linux
- **Install tools**: [kaws](https://github.com/InQuicker/kaws)
- **Others**:

**What happened**:

Created a stateful set with a persistent volume claim. Dynamic volume provisioning created an EBS volume in the us-east-1a availability zone, despite all the masters and nodes in the cluster being in us-east-1e. I tried it twice with the same results both times.

The PVC:

```
$ kubectl describe pvc -n errbit
Name:		mongodb-mongodb-0
Namespace:	errbit
StorageClass:	standard
Status:		Bound
Volume:		pvc-5fc8e90b-c8aa-11e6-9924-069508572ed2
Labels:		app=errbit
		component=mongodb
Capacity:	1Gi
Access Modes:	RWO
No events.
```

The PV created by dynamic provisioning:

```
$ kubectl describe pv
Name:		pvc-5fc8e90b-c8aa-11e6-9924-069508572ed2
Labels:		failure-domain.beta.kubernetes.io/region=us-east-1
		failure-domain.beta.kubernetes.io/zone=us-east-1a
StorageClass:	standard
Status:		Bound
Claim:		errbit/mongodb-mongodb-0
Reclaim Policy:	Delete
Access Modes:	RWO
Capacity:	1Gi
Message:
Source:
    Type:	AWSElasticBlockStore (a Persistent Disk resource in AWS)
    VolumeID:	aws://us-east-1a/vol-0d32c25dc73f029af
    FSType:	ext4
    Partition:	0
    ReadOnly:	false
No events.
```

The stateful set:

```
$ kubectl describe statefulset mongodb -n errbit
Name:			mongodb
Namespace:		errbit
Image(s):		mongo:3.4.0
Selector:		app=errbit,component=mongodb
Labels:			app=errbit,component=mongodb
Replicas:		1 current / 1 desired
Annotations:		kubectl.kubernetes.io/last-applied-configuration={"kind":"StatefulSet","apiVersion":"apps/v1beta1","metadata":{"name":"mongodb","namespace":"errbit","creationTimestamp":null,"labels":{"app":"errbit","component":"mongodb"}},"spec":{"replicas":1,"template":{"metadata":{"creationTimestamp":null,"labels":{"app":"errbit","component":"mongodb"}},"spec":{"containers":[{"name":"mongodb","image":"mongo:3.4.0","args":["--auth"],"ports":[{"name":"mongodb","containerPort":27017}],"resources":{},"volumeMounts":[{"name":"mongodb","mountPath":"/data/db"}]}]}},"volumeClaimTemplates":[{"metadata":{"name":"mongodb","creationTimestamp":null},"spec":{"accessModes":["ReadWriteOnce"],"resources":{"requests":{"storage":"1Gi"}}},"status":{}}],"serviceName":"mongodb"},"status":{"replicas":0}}
CreationTimestamp:	Thu, 22 Dec 2016 16:53:24 -0800
Pods Status:		0 Running / 1 Waiting / 0 Succeeded / 0 Failed
No volumes.
Events:
  FirstSeen	LastSeen	Count	From		SubObjectPath	Type		Reason			Message
  ---------	--------	-----	----		-------------	--------	------			-------
  12m		12m		1	{statefulset }			Normal		SuccessfulCreate	pet: mongodb-0
  11m		11m		1	{statefulset }			Normal		SuccessfulCreate	pvc: mongodb-mongodb-0
```

The stateful set's pod, pending due to the volume being in the wrong zone:

```
$ kubectl describe pod mongodb-0 -n errbit
Name:		mongodb-0
Namespace:	errbit
Node:		/
Labels:		app=errbit
		component=mongodb
Status:		Pending
IP:
Controllers:	StatefulSet/mongodb
Containers:
  mongodb:
    Image:	mongo:3.4.0
    Port:	27017/TCP
    Args:
      --auth
    Volume Mounts:
      /data/db from mongodb (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from default-token-mdqdk (ro)
    Environment Variables:	<none>
Conditions:
  Type		Status
  PodScheduled 	False
Volumes:
  mongodb:
    Type:	PersistentVolumeClaim (a reference to a PersistentVolumeClaim in the same namespace)
    ClaimName:	mongodb-mongodb-0
    ReadOnly:	false
  default-token-mdqdk:
    Type:	Secret (a volume populated by a Secret)
    SecretName:	default-token-mdqdk
QoS Class:	BestEffort
Tolerations:	<none>
Events:
  FirstSeen	LastSeen	Count	From			SubObjectPath	Type		Reason			Message
  ---------	--------	-----	----			-------------	--------	------			-------
  15m		15m		3	{default-scheduler }			Warning		FailedScheduling	[SchedulerPredicates failed due to PersistentVolume 'pvc-a5f5e714-c8a8-11e6-9924-069508572ed2' not found, which is unexpected., SchedulerPredicates failed due to PersistentVolume 'pvc-a5f5e714-c8a8-11e6-9924-069508572ed2' not found, which is unexpected.]
  14m		14m		1	{default-scheduler }			Warning		FailedScheduling	[SchedulerPredicates failed due to PersistentVolumeClaim is not bound: "mongodb-mongodb-0", which is unexpected., SchedulerPredicates failed due to PersistentVolumeClaim is not bound: "mongodb-mongodb-0", which is unexpected.]
  14m		2s		52	{default-scheduler }			Warning		FailedScheduling	pod (mongodb-0) failed to fit in any node
fit failure summary on nodes : NoVolumeZoneConflict (2)
```

The nodes, all in us-east-1e:

```
$ kubectl describe nodes | grep zone
			failure-domain.beta.kubernetes.io/zone=us-east-1e
			failure-domain.beta.kubernetes.io/zone=us-east-1e
			failure-domain.beta.kubernetes.io/zone=us-east-1e
			failure-domain.beta.kubernetes.io/zone=us-east-1e
```

**What you expected to happen**:

Dynamic volume provisioning should have created the required volume in the us-east-1e availability zone.

**How to reproduce it** (as minimally and precisely as possible):

Add the following storage class to the cluster:

``` yaml
---
kind: "StorageClass"
apiVersion: "storage.k8s.io/v1beta1"
metadata:
  name: "standard"
  annotations:
    storageclass.beta.kubernetes.io/is-default-class: "true"
provisioner: "kubernetes.io/aws-ebs"
parameters:
  type: "gp2"
  encrypted: "true"
```

Create the following stateful set and service:

``` yaml
---
kind: "Namespace"
apiVersion: "v1"
metadata:
  name: "errbit"
---
kind: "Service"
apiVersion: "v1"
metadata:
  name: "mongodb"
  namespace: "errbit"
  labels:
    app: "errbit"
    component: "mongodb"
spec:
  ports:
    - name: "mongodb"
      port: 27017
  clusterIP: "None"
  selector:
    app: "errbit"
    component: "mongodb"
---
kind: "StatefulSet"
apiVersion: "apps/v1beta1"
metadata:
  name: "mongodb"
  namespace: "errbit"
  labels:
    app: "errbit"
    component: "mongodb"
spec:
  serviceName: "mongodb"
  replicas: 1
  template:
    metadata:
      labels:
        app: "errbit"
        component: "mongodb"
    spec:
      containers:
        - name: "mongodb"
          image: "mongo:3.4.0"
          args:
            - "--auth"
          ports:
            - containerPort: 27017
              name: "mongodb"
          volumeMounts:
            - name: "mongodb"
              mountPath: "/data/db"
  volumeClaimTemplates:
    - metadata:
        name: "mongodb"
      spec:
        accessModes:
          - "ReadWriteOnce"
        resources:
          requests:
            storage: "1Gi"
```

**Comment by zachaller** [comment]
I am having this happen however we have nodes in each region and it seems to not be able to get the pod and the volume together in the same az, was setup with kubeadm with modifications to work with aws

**Comment by justinsb** [comment]
Never heard of kaws, but kubeadm is still in alpha and cloudprovider integration is not supported (i.e. expected to be broken).

You should probably be using kops to install on AWS.

**Comment by jimmycuadra** [comment]
kaws is our own installation system we've been using from the start. I don't think this bug has anything to do with the cluster creation tool. The correct cloud provider flags are passed to each Kubernetes component and other AWS-specific cloud provider functionality works.

**Comment by jimmycuadra** [comment]
If someone can point me to the area of the codebase where the decision about where to provision a dynamic volume in AWS is made, I can try to figure this out myself!

**Comment by jimmycuadra** [comment]
I think I see where the issue is. The [docs for EBS provisioning](https://github.com/kubernetes/kubernetes/blob/ef44fa1919296f67befd9436092bbcdc891b8f11/examples/experimental/persistent-volume-provisioning/#aws) say:

> `zone`: AWS zone. If not specified, a random zone from those where Kubernetes cluster has a node is chosen.

However, I have not found any logic that chooses the zone that way.

[aws.Cloud.CreateDisk](https://github.com/kubernetes/kubernetes/blob/ef44fa1919296f67befd9436092bbcdc891b8f11/pkg/cloudprovider/providers/aws/aws.go#L1580) calls its own [aws.Cloud.getAllZones](https://github.com/kubernetes/kubernetes/blob/ef44fa1919296f67befd9436092bbcdc891b8f11/pkg/cloudprovider/providers/aws/aws.go#L1039) method to populate the list of zones to choose from when creating a disk when the storage class/PVC doesn't request a specific zone. But `getAllZones` gets the zones of _all_ EC2 instances, not filtered to the Kubernetes cluster by any means at all. This list of zones is passed to [volume.util. ChooseZoneForVolume](https://github.com/kubernetes/kubernetes/blob/ef44fa1919296f67befd9436092bbcdc891b8f11/pkg/volume/util.go#L285) to pick a zone from the collection, but that function only attempts to distribute PVs across the provided zones. As such, if you have any EC2 instances running in a zone other than where your Kubernetes nodes are running, Kubernetes may pick the wrong zone.

**Comment by Dmitry1987** [comment]
+1 , happened to me too, trying to setup mongodb from a helm chart. I have a few test nodes of K8s, all in same zone, but the volumes it provisioned are in another zone, so creation of pods got stuck on 
```fit failure summary on nodes : NoVolumeZoneConflict (2), PodToleratesNodeTaints (1)```

the only way to overcome this for now is to have minions in all zones, so one of them can accept your dynamic volume with the pod?   It's still a problem on AWS... if I store a few TB of data on a volume, and trust it to be migrated during failover to another node in cluster (where failed pod will be re-created), I will be surprised to see it stuck, because K8s will try to launch a pod on any other node, with no regards to its AZ.  

But this issue is related more to cloud provider than K8s itself ...  maybe on GCE it will not happen.

**Comment by Dmitry1987** [comment]
btw, I also tried to set a default storage class, in the same availability zone where I have the nodes, with this settings:  
```
kind: StorageClass
apiVersion: storage.k8s.io/v1beta1
metadata:
  name: default
  annotations:
    storageclass.beta.kubernetes.io/is-default-class: "true"
provisioner: kubernetes.io/aws-ebs
parameters:
  type: gp2
  zone: us-west-2a
```

But the 3 volumes got created in all 3 AZs, (us-west-2a,  2b, and 2c).  Weird...  I have only this storage class, why would it create the volumes in 3 AZs even when explicitly told to use  'zone: us-west-2a' in storage class...

**Comment by jsafrane** [comment]
@jimmycuadra, you correctly found `getAllZones`, however you missed the part where it filters out all instances that are not tagged with "KubernetesCluster" tag with a specific value, it's well hidden :-).

So, tag all your AWS instances that are part of your cluster with "KubernetesCluster=jimmy" (incl. masters!) and restart Kubernetes. It should create volumes only in zones where there is an instance with the tag. You can run multiple clusters under one AWS project, as long as they have different values of KubernetesCluster tag.

@justinsb, btw, is it documented anywhere?

**Comment by jsafrane** [comment]
@Dmitry1987, that's odd, your PVs should respect parameters of storage class. Show me your PVC. It should ignore any storage class only if you use annotation `volume.alpha.kubernetes.io/storage-class` (note **alpha** there, it's important).

**Comment by Dmitry1987** [comment]
Hi @jsafrane , yes, it was *alpha* in that mongodb helm chart that I used, noticed it only after some time. thanks!

**Comment by Dmitry1987** [comment]
it worked well when i forked it and changed to "beta", then used in helm as a file, not from repo link. i have this all working now :) .

**Comment by jsafrane** [comment]
@Dmitry1987. thanks for confirmation.

@jimmycuadra, did you try KubernetesCluster tag on your AWS instances? Can we close this issue?

**Comment by jimmycuadra** [comment]
@jsafrane No, I have not confirmed that that works yet. In either case, this should be left open until the need for a KubernetesCluster tag is documented, as that is apparently critical for custom clusters. This was supposed to be tracked in #11884, but it was closed and no one has responded to my questions, which were not answered before the issue was closed.

**Comment by jimmycuadra** [comment]
I'm afraid I can't confirm that adding the `KubernetesCluster` tag with a unique value per cluster results in the behavior I'd expect. I've tagged our clusters accordingly, restarted the Kubernetes components (apiservers, controller managers, and schedulers), and created a new stateful set with the same configuration from the issue description, but Kubernetes still creates the PV in us-east-1a despite all nodes being in us-east-1e.

**Comment by jimmycuadra** [comment]
Would it be possible for the cloud provider to just use the Kubernetes labels on nodes to determine a zone, rather than using AWS API calls to try to determine which nodes should be used? It would need to look for any schedulable nodes (i.e. _not_ `--register-schedulable=false`) and look at the `failure-domain.beta.kubernetes.io/region` and `failure-domain.beta.kubernetes.io/zone` labels.

**Comment by jsafrane** [comment]
@justinsb, this idea of using Kunernetes nodes + labels instead of loading instances from AWS looks interesting to me. Could it fit into your attempt to add a caching layer to AWS provider?

**Comment by jimmycuadra** [comment]
Any updates from the storage and/or AWS teams on this? We're currently unable to use dynamic volume provisioning because of this problem.

**Comment by kingdonb** [comment]
@jimmycuadra Are you adding the label to a node of an already started cluster?

I'm not sure this will be effective.  You might have to add the label before your kube-system namespace is started for the first time.  Not sure if you're using kubeadm or something else to start your cluster, but for me, I definitely hit this issue immediately on the first try for my AWS project team that has instances in multiple AZ's.

But if I add the KubernetesCluster label before bringing the cluster up with `kubeadm --cloud-provider=aws`, I get:

```
I0301 13:41:22.736527       1 aws.go:836] AWS cloud filtering on tags: map[KubernetesCluster:kube-kingdon]
```

in my logs of kube-controller-manager-ip-172-29-151-151.ec2.internal pod.  Now my PVCs seem to be creating PVs dynamically without issue, in the right AZ every time.

**Comment by kingdonb** [comment]
Something else changes after I bring up my nodes, it might be DNS propagation delay,

When my EC2 host comes up and I've configured it using kubeadm, it's a node named `ip-172-29-151-151`.  I may `kubeadm reset` and `kubeadm init --cloud-provider=aws` again, and after some time elapses it names the node `ip-172-29-151-151.ec2.internal` instead.

This DNS name does not seem to resolve at any time before or after I notice this, I haven't been able to narrow down why this happens or what is changing, but after whatever it is, I have noticed surprises in some of the pod logs to the effect of "couldn't find the node you're talking about" with references to the new/old name.  That might be resulting in permissions errors for the backend processes that are supposed to be creating and destroying ELBs, PVs, and SGs in my AWS team.

All of this confusion has basically convinced me that I don't want to use an alpha version of kubeadm for any serious production deployment, or anything else custom, and I'll be using kops or kube-aws to build my permanent AWS deployment instead (as the official documentation recommends.)

**Comment by jimmycuadra** [comment]
We use an in-house cluster deploy tool called [kaws](https://github.com/InQuicker/kaws). I made sure to restart the kubernetes system components (e.g. controller manager) after applying the EC2 tags, so they should be visible for the cloud provider logic. Our nodes have always been named with the EC2 internal DNS. No problems with the cloud provider logic there—it's just about dynamic volume placement.

**Comment by jfoy** [comment]
Under k8s 1.5.2, we have found that tagging EC2 instances with `KubernetesCluster=foo` works for us, with one wrinkle -- we also have to ensure that we are exclusively tagging the Kubernetes controller and worker instances, instead of all instances under the VPC. (All instances are tagged by default when you set the tag via a CloudFormation option.) 

With instance tags correctly set, we see the expected behavior: PVCs using storage-provisioner aws-ebs are bound to PVs created in AZs containing cluster nodes.

**Comment by PaulFurtado** [comment]
It looks like PVC logic will also create the volume in the wrong availability zone if you specify a nodeSelector on the pod to attach it to.

There must be a better way to select the availability zone than querying the AWS API for the KubernetesCluster tag so that kubernetes pod placement logic is actually considered in the process.

**Comment by jimmycuadra** [comment]
I think using the information the API server has about the nodes is a more resilient approach. See my previous comment: https://github.com/kubernetes/kubernetes/issues/39178#issuecomment-273660193

**Comment by donjohnson** [comment]
I am seeing the same underlying behavior (full unfiltered list of zones being passed to ChooseZoneForVolume, volume placed in non-k8s zone) with GKE on Google Cloud. 

Can we treat this issue as platform-agnostic, or should I create a separate issue?

**Comment by andrewmyhre** [comment]
To add a twist to this.. I'm provisioning the Gitlab helm chart (from the [Gitlab repository](https://docs.gitlab.com/ee/install/kubernetes/gitlab_chart.html)) which provisions 3 PVCs which are used by a single pod. My nodes are in AZs us-east-1a and us-east-1c. The dynamically created PVs are each created in a random AZ (I assume either us-east-1a or us-east-1c) so the pod can only be created if all the PVs by chance happen to be created in the same AZ. Most of the time they're created in different AZs so there's no one AZ the pod can be created in which satisfies the NoVolumeZoneConflict predicate.

**Comment by rasheedamir** [comment]
just like @andrewmyhre I am provisioning jenkins which provisions two PVC's which are being used in the single pod. I have two nodes both in different AZ's. And PV's are being created randomly in different AZ's and since they are never created in same AZ; therefore my pod fails to start.

**Comment by msau42** [comment]
FYI, the issue of integrating pod scheduling with PV binding and dynamic provisioning is being tracked here #43504.  This will hopefully solve the multi-PVC In a pod issue.

However, here, there still is an issue that the AWS cloud provider library is not returning the correct zones for the cluster.

**Comment by msau42** [comment]
The current workaround for the multiple PVC zone spreading issue is to use a Statefulset, which has special zone spreading logic.

**Comment by VinceMD** [comment]
@msau42 is that documented somewhere?

**Comment by msau42** [comment]
Hm I don't think it is.  I'll see where may be the best place to add it. Maybe in the multizone page.

**Comment by msau42** [comment]
Updating docs at kubernetes/kubernetes.github.io#4706

**Comment by wangycc** [comment]
+1

**Comment by deitch** [comment]
> The current workaround for the multiple PVC zone spreading issue is to use a Statefulset, which has special zone spreading logic

Do you mind describing further @msau42 ? If I have a `StatefulSet` with `replicas: 1`, and I am working in 3 zones, how would the `volumeClaimTemplate` ensure it is in the same zone, when pod scheduling and volume scheduling remain separate?

For that matter, even if I have `replicas: 3` and 3 zones, how would I ensure that the pv for pod `-0` is in the same zone as pod `-0`, etc.? I could, theoretically, have all of the pv offset by 1 zone, and none is schedulable.

Is there a better workaround?

**Comment by msau42** [comment]
@deitch there is special zone selection logic for statefulset PVCs [here](https://github.com/kubernetes/kubernetes/blob/fc8bfe2d8929e11a898c4557f9323c482b5e8842/pkg/volume/util.go#L378).  It hashes only the statefulset name so that multiple pvcs in the same statefulset get hashed to the same value.  Then it adds the statefulset ordinal to it so that each statefulset instance will be spread across zones.

**Comment by deitch** [comment]
@msau42 so basically:

1. It keeps multiple volumes on a single pod together using the logic
2. It then allocates the PVs round-robin or whatever other logic
3. It then passes the PVs back to the pod scheduler
4. The pod scheduler then tries to allocate the pod, accounting for all of the `affinity` and `resources` and `volumes` and everything else.

But it does not (yet) take into account pod scheduling requirements when setting up the PV(s). That issue remains open, whether in a `Deployment` or a `StatefulSet`, correct? So if I have a `StatefulSet` with constraints (affinity or otherwise) that require a pod to be on a node that happens to be in zone `a`, the PV might or might not be in `a`, and so it might or might not successfully schedule the pod?

**Comment by msau42** [comment]
That is correct.  Volume provisioning today is not compatible with any sort of pod constraints that restrict the zones.

**Comment by deitch** [comment]
Got it. Any decent workarounds, beyond actually specifying the zone in the `StorageClass` (which isn't really doable because I would need to propagate it all the way through)?

**Comment by msau42** [comment]
Can you explain what you mean by propagating it all the way through?  StorageClass parameters for GCE and AWS should have a Zones parameter now, so you can specify multiple zones.

**Comment by deitch** [comment]
Sure. If I put the zone in the `StorageClass`, then I need multiple storage classes, one for each zone, e.g. `iops-east1a`, `iops-east1b`, etc.

Then, my pod (or whatever is wrapping it, deployment or statefulset) needs to be aware of which one it is calling. It no longer can just ask for 10GB of `iops`, it has to ask for 10GB of `iops-east1`. Now this config is tied specifically to `east-1a`. For that to work, I have to know that my pod for which I have some deployment affinity is in `east-1a`. For that to work, I have to constraint it to be so, so now my node config has to be aware of it.

The customization for `east-1a` propagates all the way through to lots of configs and parts.

Using `zones` in the storage class wouldn't work, because it doesn't let me ensure the right zone any more than the default (round-robin?) does. My class would need one specific zone.

And I am not even convinced this is doable with a `StatefulSet`, where there is a template, not a specific PVC. Even if I were willing to handle all of the above, how would I give each replica in the set a different `storageClassName`?

**Comment by msau42** [comment]
You don't need to explicitly constrain your pods to east-1a.  The pods will be constrained automatically by the zone where the PV is.  If you use the same spec for multiple deployments, then yes, you will need separate ones with different storage classes per zone.

For spreading your replicas across multiple zones, then you should use the `zones` parameter in the StorageClass.  That will restrict the zone spreading to only those zones you specify, instead of the default of all zones in the region.

**Comment by deitch** [comment]
> That will restrict the zone spreading to only those zones you specify, instead of the default of all zones in the region

That assumes that the node for which I have defined affinity are in those specific zones. If I have four zones, and two special nodes deployed (e.g. high memory or compute intensive or SSD), I am extremely unlikely to deploy those nodes _just_ into two special zones. I don't care which, as long as they are dispersed among zones. Now I need to know, and put that in my `StorageClass`, which is a special version of the `StorageClass` for this use case, and may require additional copies.

If I did specify, I would have to do so at the terraform (or cloudformation or whatever) level, and then pass that on to the kube configs for the `StorageClass`, which tightly ties my kube config level to my terraform level, which is quite brittle.

Make sense?

**Comment by msau42** [comment]
Yes makes sense.  Unfortunately that is why these are just workarounds until we can get the real fix in.

**Comment by deitch** [comment]
Is there a reasonable workaround for that scenario? 4 zones, 2 nodes, `Deployment` or `StatefulSet` with affinity for those nodes?

**Comment by msau42** [comment]
The only workarounds I can think of are:
- deploy your nodes across all the zones
- explicitly deploy those nodes into specific zones, and restrict the storage classes to those zones
- manually precreate the PVs in the correct zones before creating the statefulset

Multi-zonal dynamic provisioning + pod constraints on zones do not work at all right now.

**Comment by deitch** [comment]
> deploy your nodes across all the zones

with high compute nodes, it gets expensive. I only need 1-2

> explicitly deploy those nodes into specific zones, and restrict the storage classes to those zones

I might have to, but it is brittle.

> manually precreate the PVs in the correct zones before creating the statefulset

Not necessarily any easier than the second option, and still requires work outside the cluster.

Oh well. Thanks for walking it through.

**Comment by boeboe** [comment]
+1

**Comment by boeboe** [comment]
@msau42 this would work if a POD only has 1 PVC. In my case I have a POD with 2 PVC's and then this trick doesn't work.

**Comment by justinsb** [comment]
@deitch can you explain what you're trying to do?

On AWS, the cluster autoscaler comes into play as well - instances are dynamic.  So what I expect the answer to be:

* If you're allocating in all N zones, cluster autoscaler _should_ create instances to ensure that pods are schedulable.
* If you're allocating in a subset of zones, you would specify a storage class constraining to those zones, I think the statefulset logic picks up on that today, and then a cluster autoscaler _should_ creates instances in just those zones.
* Likely you use taints to enforce that only your particular workloads end up on your "expensive" nodes, and not random low-priority workloads.

The root challenge here is that if you want a subset of zones, that volumes can't be moved between zones (today), so once you create a volume, it's stuck there.  Unless you constrain the set of zones in which your special instances live, eventually you'll be in all zones.  A storage class should do that today.

I _think_ cluster autoscaler does most of this today if you taint your nodes.   I don't know if cluster autoscaler infers the zones from volumes when considering unschedulable pods.

I think in practice if you create a kops InstanceGroup (or whatever the equivalent is that maps to ASGs / MIGs), and make sure that your InstanceGroup is tainted and constrained to two zones that match the zones in the StorageClass, that it'll work _ok_ today.  Even if the autoscaler isn't zone-aware, we can assume that AWS will place instances roughly balanced between the zones, and that your workload will be roughly symmetrical across the zones because of the StatefulSet spreading logic.  Not perfect, but should work OK today in practice.

Is this the gap @deitch - that the cluster autoscaler won't necessarily create new nodes in the correct zones?  (And if not, what's missing?)  If so, this feels like a relatively easy fix (tweak cluster autoscaler to be zone aware), with a workaround today (create a dedicated ASG tied to the zones you're using).

**Comment by deitch** [comment]
@justinsb :

> Likely you use taints to enforce that only your particular workloads end up on your "expensive" nodes, and not random low-priority workloads

Precisely.

> The root challenge here is that if you want a subset of zones, that volumes can't be moved between zones (today), so once you create a volume, it's stuck there

Definitely. Or perhaps slightly more accurately, "if you end up in a subset of zones for reasons of choice or constraints". I think if EFS performance were to match EBS, especially cross-AZ (haven't tested), and the EFS provisioner were to be promoted, then a lot of this would go away.

But there also is an issue even if I have not constrained my specific workers to specific zones. If I have regular workers in 3 zones, but only launch 1 tainted worker, then it is, by definition, only in 1 zone. The volume "scheduler" does _not_ take the pod scheduler into account. It just picks a zone, creates the volume and then hands its info off to the pod scheduler, which now cannot schedule it.

IMHO, there are several ways to resolve this:

* Make the volume "scheduler" pod constraint (topology) aware
* Make the volume "re-zonable" (which is an AWS / cloud provider issue, nothing k8s can do about it)

FYI, I am not using the cluster auto-scaler here. I just created 1 tainted node in the cluster; AWS picks a zone and places it. 

I think that as long as a volume is not "re-zonable", then the volume scheduler should know which zones a pod could go into (or on which nodes, and therefore which zones), prior to creating it.

I do admit that, in general, volumes will be an issue. A 3-zone cluster, with a 3-replica `StatefulSet` with attached PVs, that loses a zone or a single node in a zone but no other node in that zone meets the `StatefulSet` resource or affinity constraints, will have the pod unschedulable until a node that matches those constraints in that zone becomes available.

But at least I can get started if the volume is pod-scheduler aware.

**Comment by justinsb** [comment]
@deitch - what you're saying makes sense.  I think the challenge is that there's another use case as well - the clustered DB case.  If you have a 3 node DB and 3 AZs, we want to allocate those volumes in all 3 AZs.  Now the controversial case: I think we want to do that even if one of the AZs is down right now, so the pod won't be scheduled - eventually it'll recover, and you don't want your DB always to be in 2 AZs because one of the AZs happened to be down when you created it.

I think the current heuristics work well when you have lots of StatefulSets that each cover most of the AZs.  They work less well when we get to the "one node" or "one AZ" case.

But ... what about setting a zone in the Zones field of your StorageClass, and launching your instances into that zone?  We need some way for you to indicate that you are in the "focus on these zones", rather than the "spread across zones for availability" case, and the Zones field of the StorageClass seems reasonable.

**Comment by deitch** [comment]
@justinsb - I agree with the challenge. In the case of no volume (or no zone dependency), then if I have a "preferredWhenScheduling"-type affinity, then when the new zone comes up, it eventually will rebalance into all 3 AZ. If there is a volume involved, then it gets messy.

I think the way I would handle it (in the absence of zone-migratable or multi-zone-accessible volumes) is to set an affinity that prevents having more than one instance in one zone, so if I only have 2 zones, I will start with 2 instances, growing to 3 when a third becomes available. Essentially "requiredWhenScheduling"-type affinity.

> what about setting a zone in the Zones field of your StorageClass, and launching your instances into that zone? 

That would work, but then I have to specify an actual zone for the node, and taint not just a class of nodes but class and zone, and know it when scheduling the pod, and my pvc has to match it, and statefulsets (with a template) would be difficult to do. It creates a lot of brittleness that goes against the spirit of kube.

Even if I do actually want _this_ node in _this_ zone (in most cases, it just happens that way), then I would prefer to have a taint, tolerate that taint on my pod, and then let the scheduling of pod and volumes "just work."

**Comment by lmpetrie13** [comment]
Any updates on this issue? 

I'm dealing with a similar situation... I've got 3 nodes, spread across 2 zones. I've then got a StatefulSet which deploys 3 pods, that have an antiAffinity with each other(so 1 per node). The StatefulSet also provisions one persistent volume per pod. So far every attempt to deploy has resulted in the following Scenario(Or it's reverse)...

Zone A: Node1, Node2, Pod1, Pod2, Volume1
Zone B: Node3, Pod3, Volume2, Volume3

The ratio of Volumes to Pods has been wrong every time which results in a failure(NoVolumeZoneConflict).

Aside from making all the nodes and volumes spin up in just one zone(which defeats the purpose) I've yet to think of/find a solution.

**Comment by msau42** [comment]
There's 2 problems being discussed in this issue:
1) OP, on AWS, volumes are getting provisioned in zones where nodes do not exist.  Let's use this issue to discuss this.
2) Dynamic provisioning is not smart enough to look at pod affinity/anti-affinity policies.  Let's use https://github.com/kubernetes/kubernetes/issues/49906 to discuss that.  Solving that issue is currently under design.

**Comment by msau42** [comment]
@gnufied @jsafrane has the first issue been fixed for aws?

**Comment by StephanX** [comment]
@msau42 In my experience, it isn't simply that volumes are provisioned in availability zones where nodes do not exist, but also that pods are scheduled independently from PVs (on statefulset creation), then PVs are created without regard for scheduled pod locations.  Later, if a pod is reaped, there's no guarantee that it will be rescheduled in an AZ that matches the existing PV. 

My best effort work around has been to create custom storage classes to pin the statefulset to a zone, which negates the benefits of a cluster that spans multiple AZs.

**Comment by msau42** [comment]
@StephanX agree, but because the solutions for the two are completely different, I want to split them out into separate issues and track them separately.

**Comment by ankurgadgilwar** [comment]
Hello all, 
I am pretty new to Kubernetes. What I am trying to do is get a VolumeClaimTemplate to attach to a storage class that i have defined already. 
The question I have is, would the claim bind itself to the StorageClass mentioned in the YAML file if I haven't defined the volumes part before the volumeClaimTemplate part?
`volumeMounts:
          name: zk-datadir
          annotations:
            volume.alpha.kubernetes.io/storage-class: gold
        - mountPath: /var/lib/zookeeper
  volumeClaimTemplates:
  - metadata:
      name: zk-datadir
    spec:
      accessModes: [ "ReadWriteOnce" ]
      resources:
        requests:
          storage: 1Gi`
Will this work if i havent defined the volume?

**Comment by zonepalms91** [comment]
I'm seeing this issue as well, but the only workaround when dealing with large clusters across all AZs is to use an EFS mount, which is less than ideal.

**Comment by piraluc** [comment]
Just in case somebody wants to cross-check: After removing the `-` from the name of my release it worked. So instead of `helm install ./foo --name foo-bar --namespace foo` I used `helm install ./foo --name foobar --namespace foo`.

**Comment by msau42** [comment]
@edwardsmit  Topology aware volume provisioning in 1.12 should help with provisioning volumes in zones that can meet your Pod scheduling requirements.

**Comment by edwardsmit** [comment]
Thank you for the response @msau42, do you know of an issue I can follow?

**Comment by msau42** [comment]
Feature issue is here: https://github.com/kubernetes/enhancements/issues/490
1.12 blog post with examples is here: https://kubernetes.io/blog/2018/10/11/topology-aware-volume-provisioning-in-kubernetes/
Official documentation is here: https://kubernetes.io/docs/concepts/storage/storage-classes/#volume-binding-mode

**Comment by lig** [comment]
@msau42 this one looks resolved

**Comment by bartelsb** [comment]
@msau42 Does the new feature handle the case where a node with an existing pod/volume goes down and the pod gets moved to a node in a different AZ? Specifically, does Kubernetes handle copying the existing volume to a new volume in the new AZ? If not, when the pod gets redeployed in a new AZ, it would no longer have access to any data from the old volume.

I reviewed both the blog post and the official documentation, but I didn't see anything that addressed this specific case. Thanks!

**Comment by msau42** [comment]
No, data migration is not part of this feature.  The feature only handles initial provisioning of a volume.  Once the volume is provisioned, it must always be scheduled to a node in the same zone.

If you need to handle zone outages, you will need to use a storage system that does cross zone replication.

**Comment by silashansen** [comment]
We have different node-types for different workloads. This is accomplished by using node taints.
Some node-types run in multiple zones and others don't.
Since the unique list of zones for the entire cluster make up the list of possible zones for the PV, we ended up with PV's in an zone that didn't have any nodes that our workload could be scheduled on.

Only looking at zones for nodes that where the pod can actually be scheduled, when selecting the zone for the PV, would be great!

**Comment by RobertFr0st** [comment]
This appears to me to still be an ongoing issue.

**Comment by Efrat19** [comment]
still happens on EKS 1.21

**Comment by cprak-nydig** [comment]
still running into this issue on 1.21.5

**Comment by hfuss** [comment]
So the issue has been solved as mentioned in https://github.com/kubernetes/kubernetes/issues/39178#issuecomment-436033217

You need to have a `StorageClass` which uses `WaitForFirstConsumer` as the `bindingMode`. Otherwise with the `Immediate` mode that most default classes use, the volume provisioner will select any AZ which might not align with where the consumer of the volume ends up getting scheduled.
